WIP: Return NVME formatted UUID when volume is attached to VM on NVMe controller

chethanv28 opened the pull request:
<!-- Thanks for sending a pull request! -->

**What this PR does / why we need it**:
This PR is a draft PR for returning NVMe formatted UUID for volumes attached to VMs on NVMe controller
TODO: Add supervisor cluster flavor check for the new code the PR is adding

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Testing done**:
To be added

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Return NVME formatted UUID when volume is attached to VM on NVMe controller
```

chethanv28 commented:
jtest block-vanilla
jtest wcp

svcbot-qecnsdp commented:
Started Vanilla block pipeline...

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
``` 
Block vanilla build status: FAILURE 
Stage before exit: testbed-deploy 

```

svcbot-qecnsdp commented:
Started Vanilla block pipeline...

svcbot-qecnsdp commented:
Started Vanilla block pipeline...

svcbot-qecnsdp commented:
Started Vanilla block pipeline...

chethanv28 commented on pkg/common/cns-lib/volume/manager.go:
```diff
@@ -56,7 +56,7 @@ type Manager interface {
 	CreateVolume(ctx context.Context, spec *cnstypes.CnsVolumeCreateSpec) (*CnsVolumeInfo, error)
 	// AttachVolume attaches a volume to a virtual machine given the spec.
-	AttachVolume(ctx context.Context, vm *cnsvsphere.VirtualMachine, volumeID string) (string, error)
+	AttachVolume(ctx context.Context, vm *cnsvsphere.VirtualMachine, volumeID string, checkNVMeController bool) (string, error)
```
Introducing this bool to make sure NVMe controller check is only for supervisor cluster

RaunakShah replied:
How about adding this as a parameter to `defaultManager`? You'd just need to pass this parameter once during init based on cluster flavor.

chethanv28 replied:
I will follow-up with this suggestion in the next iteration. Created an issue to track the same: https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/913

RaunakShah commented on pkg/common/cns-lib/volume/util.go:
```diff
@@ -66,11 +75,49 @@ func IsDiskAttached(ctx context.Context, vm *cnsvsphere.VirtualMachine, volumeID
+		return "", err
+	}
+	var nvmeUUID uuidlib.UUID
+	nvmeUUID[0] = uuidBytes[8]
```
Is the difference little endian vs big endian ? Is there a golang package that can do that conversion?

zhelongp replied:
> Is there a golang package that can do that conversion?
It would be better to look for packages only from wellknown places (e.g. golang standard libraries or k8s packages).
Minimizing external dependency will make your code simpler.

chethanv28 replied:
Let me take this an action item to see if there are well known packages to replace the conversion logic

chethanv28 replied:
I will follow-up with this suggestion in the next iteration. Created an issue to track the same: https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/913

divyenpatel commented on pkg/common/cns-lib/volume/util.go:
```diff
@@ -56,6 +57,14 @@ func IsDiskAttached(ctx context.Context, vm *cnsvsphere.VirtualMachine, volumeID
 					virtualDevice := device.GetVirtualDevice()
 					if backing, ok := virtualDevice.Backing.(*vimtypes.VirtualDiskFlatVer2BackingInfo); ok {
 						log.Infof("Found diskUUID %s for volume %s on vm %+v", backing.Uuid, volumeID, vm)
+						if checkNVMeController && hasNVMEController(ctx, vmDevices) {
```
what if the disk is not attached to the NVME controller and attached non-NVME controller.
We should not assume if VM has NVME controller, then disk is attached to that NVME controller,.

chethanv28 replied:
Addressed by checking if the disk on the NVMe controller

divyenpatel reviewed (changes requested):
Can you address 

- https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/890/files#r637102861 
- https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/890/files#r637113030

divyenpatel commented on pkg/common/cns-lib/volume/util.go:
```diff
@@ -66,11 +75,49 @@ func IsDiskAttached(ctx context.Context, vm *cnsvsphere.VirtualMachine, volumeID
 
+func hasNVMEController(ctx context.Context, devices object.VirtualDeviceList) bool {
+	log := logger.GetLogger(ctx)
+	_, err := devices.FindNVMEController("")
```
`devices.FindNVMEController` may return nil if no NVMEController is available with the free slots.
For the last disk attached to NVME controller, if we get a fault and call this function, we will get nil, so we will not be able to return success for already attached disk to NVME controller.


`devices.FindNVMEController` is internally using `PickController`

```golang
	c := l.PickController((*types.VirtualNVMEController)(nil))
	if c == nil {
		return nil, errors.New("no available NVME controller")
	}
```

```golang
// PickController returns a controller of the given type(s).
// If no controllers are found or have no available slots, then nil is returned.
func (l VirtualDeviceList) PickController(kind types.BaseVirtualController) types.BaseVirtualController {
```

chethanv28 replied:
I removed this function call: `FindNVMEController`. We need to check the controllerKey on the disk and see if the controllerKey matches with the NVMe controller key

chethanv28 commented:
jtest block-vanilla

svcbot-qecnsdp commented:
Started Vanilla block pipeline...

chethanv28 commented:
jtest wcp

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
``` 
WCP build status: FAILURE 
Stage before exit: null 
Jenkins E2E Test Results: 
Ran 32 of 179 Specs in 7263.866 seconds
FAIL! -- 1 Passed | 31 Failed | 0 Pending | 147 Skipped
--- FAIL: TestE2E (7264.03s)
FAIL

Ginkgo ran 1 suite in 2h1m25.273663977s
Test Suite Failed
make: Leaving directory `/home/worker/workspace/github-csi-wcp-CICD/Results/655/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
``` 
Block vanilla build status: FAILURE 
Stage before exit: e2e-tests 
Jenkins E2E Test Results: 
Ran 43 of 179 Specs in 10764.383 seconds
FAIL! -- 38 Passed | 5 Failed | 0 Pending | 136 Skipped
--- FAIL: TestE2E (10764.47s)
FAIL

Ginkgo ran 1 suite in 2h59m46.25421022s
Test Suite Failed
make: Leaving directory `/home/worker/workspace/github-csi-block-vanilla@3/Results/551/vsphere-csi-driver`

```

SandeepPissay reviewed (commented):
> TODO: Add supervisor cluster flavor check for the new code the PR is adding

Is this PR ready for review? If no, why do you have a TODO?

SandeepPissay commented:
@chethanv28 looks like you do not have a successful pipelines run. Can you run the E2E tests to make sure that this PR is tested well?

SandeepPissay commented on pkg/common/cns-lib/volume/util.go:
```diff
@@ -41,36 +42,92 @@ func validateManager(ctx context.Context, m *defaultManager) error {
 				if virtualDisk.VDiskId != nil && virtualDisk.VDiskId.Id == volumeID {
 					virtualDevice := device.GetVirtualDevice()
-					if backing, ok := virtualDevice.Backing.(*vimtypes.VirtualDiskFlatVer2BackingInfo); ok {
+					if value, ok := nvmeControllerKeyToNameMap[virtualDevice.ControllerKey]; ok {
```
I have not looked at a VM config info with NVMe controller, so I'm not sure on the logic followed here. Could you please validate the logic with Sriram?

chethanv28 replied:
This logic was validated with Sriram & ST and all the tests passed

SandeepPissay commented on pkg/common/cns-lib/volume/util.go:
```diff
@@ -41,36 +42,92 @@ func validateManager(ctx context.Context, m *defaultManager) error {
 					virtualDevice := device.GetVirtualDevice()
-					if backing, ok := virtualDevice.Backing.(*vimtypes.VirtualDiskFlatVer2BackingInfo); ok {
+					if value, ok := nvmeControllerKeyToNameMap[virtualDevice.ControllerKey]; ok {
+						log.Info("Found that the disk in contention is attached to NVMe controller")
```
Instead of this static log message, how about this:

`Found disk %s is attached to NVMe controller on VM %s.`

And could you please move the log to debug?

chethanv28 replied:
Done

SandeepPissay commented on pkg/common/cns-lib/volume/util.go:
```diff
@@ -41,36 +42,92 @@ func validateManager(ctx context.Context, m *defaultManager) error {
+					if value, ok := nvmeControllerKeyToNameMap[virtualDevice.ControllerKey]; ok {
+						log.Info("Found that the disk in contention is attached to NVMe controller")
+						if strings.Contains(value, "NVME") {
+							log.Info("Fetching the backing uuid for the disk on NVMe controller")
```
You may want to improve this static message too. And please move it to debug.

chethanv28 replied:
Removed this log since it was only needed for debugging on ST setups

SandeepPissay commented on pkg/common/cns-lib/volume/util.go:
```diff
@@ -41,36 +42,92 @@ func validateManager(ctx context.Context, m *defaultManager) error {
+							}
+						}
+					}
+					log.Info("Disk in contention is not attached to NVMe controller")
```
Redundant log?

chethanv28 replied:
Removed. Used this to verify logs on ST runs

SandeepPissay commented on pkg/common/cns-lib/volume/util.go:
```diff
@@ -41,36 +42,92 @@ func validateManager(ctx context.Context, m *defaultManager) error {
+									log.Errorf("failed to convert uuid to  NvmeV13UUID for the vm: %s", vm.InventoryPath)
+									return "", err
+								}
+								log.Infof("Successfully converted diskUUID %s to NvmeV13UUID %s for volume %s on vm %+v", backing.Uuid, uuid, volumeID, vm)
```
Can be moved to debug.

chethanv28 replied:
Done

SandeepPissay commented on pkg/common/cns-lib/volume/util.go:
```diff
@@ -41,36 +42,92 @@ func validateManager(ctx context.Context, m *defaultManager) error {
+					}
+					log.Info("Disk in contention is not attached to NVMe controller")
+					if backing, ok := virtualDevice.Backing.(*types.VirtualDiskFlatVer2BackingInfo); ok {
 						log.Infof("Found diskUUID %s for volume %s on vm %+v", backing.Uuid, volumeID, vm)
```
Can be moved to debug log.

chethanv28 replied:
I would let this log be Infof, since I made sure all other logs are debug. This is the only log which will help us debug in prod to verify if disk is attached to scsi or nvme controller

SandeepPissay commented on pkg/common/cns-lib/volume/util.go:
```diff
@@ -41,36 +42,92 @@ func validateManager(ctx context.Context, m *defaultManager) error {
 // IsDiskAttached checks if the volume is attached to the VM.
 // If the volume is attached to the VM, return disk uuid of the volume, else return empty string
-func IsDiskAttached(ctx context.Context, vm *cnsvsphere.VirtualMachine, volumeID string) (string, error) {
+func IsDiskAttached(ctx context.Context, vm *cnsvsphere.VirtualMachine, volumeID string, checkNVMeController bool) (string, error) {
```
Where is `checkNVMeController` used?

chethanv28 commented:
> > TODO: Add supervisor cluster flavor check for the new code the PR is adding
> 
> Is this PR ready for review? If no, why do you have a TODO?

My bad, I had added this when I published the PR since I wanted to run e2e tests first. Removed the TODO

SandeepPissay commented on pkg/common/cns-lib/volume/util.go:
```diff
@@ -41,36 +42,92 @@ func validateManager(ctx context.Context, m *defaultManager) error {
 				if virtualDisk.VDiskId != nil && virtualDisk.VDiskId.Id == volumeID {
 					virtualDevice := device.GetVirtualDevice()
-					if backing, ok := virtualDevice.Backing.(*vimtypes.VirtualDiskFlatVer2BackingInfo); ok {
+					if checkNVMeController {
```
Why do we need this check? I think removing this check will work for disk attached to NVMe controller as well as PVSCSI controller. Am I missing something?

chethanv28 replied:
It will work, but I using this check as a guard for vanilla flavor. Is it ok to use the conversion logic for vanilla flavor as well? If yes, then I can remove the `checkNVMeController`

SandeepPissay replied:
Lets not enable NVMe related logic to vanilla cluster.

chethanv28 replied:
ok, I will keep the checkNVMeController check & set it to false for vanilla k8s

chethanv28 commented:
> @chethanv28 looks like you do not have a successful pipelines run. Can you run the E2E tests to make sure that this PR is tested well?

I see there are lot failure runs on WCP pipeline and it is failing in beforeEach of all the test suites. Seems like a core e2e framework issue. Created bugs for these failures.

svcbot-qecnsdp commented:
Started WCP block pipeline...

divyenpatel reviewed (approved):
/approve

divyenpatel commented:
@chethanv28 can you squash all your commits?

chethanv28 commented:
> @chethanv28 can you squash all your commits?

Done

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
``` 
WCP build status: FAILURE 
Stage before exit: null 

```

svcbot-qecnsdp commented:
``` 
WCP build status: SUCCESS 
Stage before exit: finally 
Jenkins E2E Test Results: 

Ran 32 of 186 Specs in 7102.870 seconds
SUCCESS! -- 32 Passed | 0 Failed | 0 Pending | 154 Skipped
PASS

Ginkgo ran 1 suite in 1h58m43.932214489s
Test Suite Passed
make: Leaving directory `/home/worker/workspace/github-csi-wcp-CICD@4/Results/666/vsphere-csi-driver`

```

chethanv28 commented:
> ```
> WCP build status: SUCCESS 
> Stage before exit: finally 
> Jenkins E2E Test Results: 
> 
> Ran 32 of 186 Specs in 7102.870 seconds
> SUCCESS! -- 32 Passed | 0 Failed | 0 Pending | 154 Skipped
> PASS
> 
> Ginkgo ran 1 suite in 1h58m43.932214489s
> Test Suite Passed
> make: Leaving directory `/home/worker/workspace/github-csi-wcp-CICD@4/Results/666/vsphere-csi-driver`
> ```

@SandeepPissay @divyenpatel Can you help with approval and merging of the PR

SandeepPissay commented:
I'm capturing one of the concern we discussed. Basically this code is enabled to return NVME formatted UUID only for supervisor cluster. We discussed that a vanilla k8s cluster can be created with vNVMe controller and CNS would attach the volume to this controller without knowing whether the VM is a Pod VM or a regular vanilla node VM. But, the vanilla CSI does not expect the volume to be attached to vNVMe controller and the CSI would break. So, its not clear how do we expect things to work in such scenarios. We should resolve this soon.

chethanv28 commented:
> I'm capturing one of the concern we discussed. Basically this code is enabled to return NVME formatted UUID only for supervisor cluster. We discussed that a vanilla k8s cluster can be created with vNVMe controller and CNS would attach the volume to this controller without knowing whether the VM is a Pod VM or a regular vanilla node VM. But, the vanilla CSI does not expect the volume to be attached to vNVMe controller and the CSI would break. So, its not clear how do we expect things to work in such scenarios. We should resolve this soon.

Is there any resolution needed for CSI code in vanilla k8s because of this change ?

SandeepPissay commented:
> > I'm capturing one of the concern we discussed. Basically this code is enabled to return NVME formatted UUID only for supervisor cluster. We discussed that a vanilla k8s cluster can be created with vNVMe controller and CNS would attach the volume to this controller without knowing whether the VM is a Pod VM or a regular vanilla node VM. But, the vanilla CSI does not expect the volume to be attached to vNVMe controller and the CSI would break. So, its not clear how do we expect things to work in such scenarios. We should resolve this soon.
> 
> Is there any resolution needed for CSI code in vanilla k8s because of this change ?

Yes, vanilla CSI would break if the vanilla node VM has at least one NVMe controller.

divyenpatel commented:
/hold

svcbot-qecnsdp commented:
Started Vanilla block pipeline...

svcbot-qecnsdp commented:
``` 
Block vanilla build status: FAILURE 
Stage before exit: testbed-deploy 

```

svcbot-qecnsdp commented:
``` 
Block vanilla build status: FAILURE 
Stage before exit: checkout 

```

svcbot-qecnsdp commented:
Started Vanilla block pipeline...

svcbot-qecnsdp commented:
``` 
Block vanilla build status: SUCCESS 
Stage before exit: finally 
Jenkins E2E Test Results: 

Ran 1 of 186 Specs in 9.716 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 185 Skipped
PASS

Ginkgo ran 1 suite in 52.005366243s
Test Suite Passed
make: Leaving directory `/home/worker/workspace/github-csi-block-vanilla/Results/570/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
Started Vanilla block pipeline...

chethanv28 commented:
> ```
> Block vanilla build status: SUCCESS 
> Stage before exit: finally 
> Jenkins E2E Test Results: 
> 
> Ran 1 of 186 Specs in 9.716 seconds
> SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 185 Skipped
> PASS
> 
> Ginkgo ran 1 suite in 52.005366243s
> Test Suite Passed
> make: Leaving directory `/home/worker/workspace/github-csi-block-vanilla/Results/570/vsphere-csi-driver`
> ```

This was a dry run to test my changes using e2e pipeline with an existing testbed. Ran one simple volume provisioning test.
Ignore this.
Will monitor this run : https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/890#issuecomment-848296786

divyenpatel commented:
/hold cancel

divyenpatel reviewed (approved):
/approve

divyenpatel commented:
/ok-to-test

chethanv28 commented:
> > > I'm capturing one of the concern we discussed. Basically this code is enabled to return NVME formatted UUID only for supervisor cluster. We discussed that a vanilla k8s cluster can be created with vNVMe controller and CNS would attach the volume to this controller without knowing whether the VM is a Pod VM or a regular vanilla node VM. But, the vanilla CSI does not expect the volume to be attached to vNVMe controller and the CSI would break. So, its not clear how do we expect things to work in such scenarios. We should resolve this soon.
> > 
> > 
> > Is there any resolution needed for CSI code in vanilla k8s because of this change ?
> 
> Yes, vanilla CSI would break if the vanilla node VM has at least one NVMe controller.

As discussed, the disk attached to nvme controllers is handled only for WCP podVMs in the CNS layer as well. And, this PR is adding code to return nvme style uuid only for WCP flavors. I think this is now resolved.

svcbot-qecnsdp commented:
``` 
Block vanilla build status: FAILURE 
Stage before exit: e2e-tests 
Jenkins E2E Test Results: 
Ran 43 of 186 Specs in 10652.630 seconds
FAIL! -- 39 Passed | 4 Failed | 0 Pending | 143 Skipped
--- FAIL: TestE2E (10652.72s)
FAIL

Ginkgo ran 1 suite in 2h58m25.506717672s
Test Suite Failed
make: Leaving directory `/home/worker/workspace/github-csi-block-vanilla/Results/571/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
Started Vanilla block pipeline...

svcbot-qecnsdp commented:
``` 
Block vanilla build status: SUCCESS 
Stage before exit: finally 
Jenkins E2E Test Results: 

Ran 1 of 186 Specs in 351.733 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 185 Skipped
PASS

Ginkgo ran 1 suite in 6m27.524112644s
Test Suite Passed
make: Leaving directory `/home/worker/workspace/github-csi-block-vanilla@2/Results/575/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
Started Vanilla block pipeline...

svcbot-qecnsdp commented:
``` 
Block vanilla build status: SUCCESS 
Stage before exit: finally 
Jenkins E2E Test Results: 

Ran 1 of 186 Specs in 420.177 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 185 Skipped
PASS

Ginkgo ran 1 suite in 7m39.149760314s
Test Suite Passed
make: Leaving directory `/home/worker/workspace/github-csi-block-vanilla@2/Results/576/vsphere-csi-driver`

```

chethanv28 commented:
> ```
> Block vanilla build status: FAILURE 
> Stage before exit: e2e-tests 
> Jenkins E2E Test Results: 
> Ran 43 of 186 Specs in 10652.630 seconds
> FAIL! -- 39 Passed | 4 Failed | 0 Pending | 143 Skipped
> --- FAIL: TestE2E (10652.72s)
> FAIL
> 
> Ginkgo ran 1 suite in 2h58m25.506717672s
> Test Suite Failed
> make: Leaving directory `/home/worker/workspace/github-csi-block-vanilla/Results/571/vsphere-csi-driver`
> ```

Summarizing the 4 failures not related to this PR:
```
[Fail] [csi-block-vanilla] [csi-file-vanilla] CNS-CSI Cluster Distribution Operations during VC reboot [It] [csi-block-vanilla] [csi-file-vanilla] verify volume operations after vc reboots 
/home/worker/workspace/github-csi-block-vanilla/Results/571/vsphere-csi-driver/tests/e2e/csi_cns_telemetry_vc_reboot.go:361

[Fail] [csi-block-vanilla] full-sync-test [It] Verify PVC metadata is deleted in CNS after PVC is deleted in k8s 
/home/worker/workspace/github-csi-block-vanilla/Results/571/vsphere-csi-driver/tests/e2e/fullsync_test_for_block_volume.go:627

[Fail] [csi-block-vanilla] full-sync-test [It] Verify Multiple PVCs are deleted/updated after full sync 
/home/worker/workspace/github-csi-block-vanilla/Results/571/vsphere-csi-driver/tests/e2e/fullsync_test_for_block_volume.go:444

[Fail] [csi-block-vanilla] [csi-file-vanilla] CNS-CSI Cluster Distribution Telemetry [It] Verify dynamic provisioning of pvc has cluster-distribution value updated 
/home/worker/workspace/github-csi-block-vanilla/Results/571/vsphere-csi-driver/tests/e2e/csi_cns_telemetry.go:131
```

2 full sync tests failed due to a known bug which was fixed this morning. I re-ran the tests after rebasing the PR here: https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/890#issuecomment-848442625 & https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/890#issuecomment-848450363

2 CNS-CSI Cluster distribution telemetry tests are failing intermittently on the pipeline due to a known bug. They are not related to this PR as well

SandeepPissay commented:
@chethanv28 Its a good practice to close on open comments in general. Can you take care of this in the future? Btw, I'm fine with the current code changes and I also see that the E2E tests are run with some failures (according to Chethan they are unrelated to the PR). Can you file P0 bugs on the failed tests?

/lgtm

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
``` 
Build ID: 4
WCP build status: FAILURE 
Stage before exit: null 
Jenkins E2E Test Results: 
Ran 14 of 186 Specs in 3277.798 seconds
FAIL! -- 10 Passed | 4 Failed | 0 Pending | 172 Skipped
--- FAIL: TestE2E (3277.97s)
FAIL

Ginkgo ran 1 suite in 54m52.419842622s
Test Suite Failed
make: Leaving directory `/home/worker/workspace/csi-wcp-pre-check-in@2/Results/4/vsphere-csi-driver`

```
